tools/AzureRM.BootStrapper/AzureRM.Bootstrapper.psm1

```diff
@@ -66,38 +146,478 @@ function Get-AzProfile
+    if($Profile -notin $PreviousProfiles)
+    {
+      return
+    }
+
+    if ($Modules -eq $null)
+    {
+      $Modules = ($PreviousMap.$Profile | Get-Member -MemberType NoteProperty).Name
+    }
+
+    foreach ($module in $Modules)
+    {
+      # If the latest version is same as the previous version, do not uninstall.
+      if (($PreviousMap.$Profile.$module | Out-String) -eq ($LatestMap.$Profile.$module | Out-String))
+      {
+        continue
+      }
+      
+      # Is that module installed? If not skip
+      $versionList = $PreviousMap.$Profile.$module
+      foreach ($version in $versionList)
+      {
+        if ((Get-Module -Name $Module -ListAvailable | Where-Object { $_.Version -eq $version} ) -eq $null)
+        {
+          continue
+        }
+
+        # Modules are different. Uninstall previous version.
+        if ($Remove.IsPresent)
+        {
+          Invoke-UninstallModule -PMap $PreviousMap -Profile $Profile -Module $module -AllProfilesInstalled $AllProfilesInstalled -RemovePreviousVersions
+        }
+        else {
+          Invoke-UninstallModule -PMap $PreviousMap -Profile $Profile -Module $module -AllProfilesInstalled $AllProfilesInstalled         
+        }
+      }
+    }
+  }
+}
+
+# Gets profiles installed from all the profilemaps from cache
+function Get-AllProfilesInstalled
+{
+  $AllProfilesInstalled = @{}
+  $ProfileCache = Get-ProfileCachePath
+  if (-not (Test-Path $ProfileCache))
+  {
+    return
+  }
+  $ProfileMapHashes = Get-ChildItem $ProfileCache 
+  foreach ($ProfileMapHash in $ProfileMapHashes)
+  {
+    $ProfileMap = Get-Content -Raw -Path (Join-Path $ProfileCache $ProfileMapHash.Name) |  ConvertFrom-json
+    $profilesInstalled = (Get-ProfilesInstalled -ProfileMap $ProfileMap)
+    foreach ($Profile in $profilesInstalled.Keys)
+    {
+      foreach ($Module in ($ProfileMap.$Profile | Get-Member -MemberType NoteProperty).Name)
+      {
+        $versionList = $ProfileMap.$Profile.$Module
+        foreach ($version in $versionList)
+        {
+          if ($AllProfilesInstalled.ContainsKey(($Module + $version)))
+          {
+            if ($Profile -notin $AllProfilesInstalled[($Module + $version)])
+            {
+              $AllProfilesInstalled[($Module + $version)] += $Profile
+            }
+          }
+          else {
+            $AllProfilesInstalled.Add(($Module + $version), @($Profile))
+          }
+        }
+      }
+    }
+  }
+  return $AllProfilesInstalled
+}
+
+# Helps to remove-previous versions of the update-profile and clean up cache, if none of the old hash profiles are installed
+function Update-ProfileHelper
+{
+  [CmdletBinding()]
+  param()
+  DynamicParam
+  {
+    $params = New-Object -Type System.Management.Automation.RuntimeDefinedParameterDictionary
+    Add-ProfileParam $params
+    Add-RemoveParam $params
+    Add-ModuleParam $params
+    return $params
+  }
+  
+  PROCESS
+  {
+    $Profile = $PSBoundParameters.Profile
```

Comment:
This comment was discussed offline. The locks this comment mentions was on profile cache. Profilemap will not have read/write simultaneously since we always would create a new profile map if update is available online. We have retries for reading and writing as well in case of simultaneous access requests.